### ISSUE TITLE
Can't compile when proto definition contains oneOf field type

### ISSUE THREAD

**Issue by donpaul120** [grpc/grpc-java#6306]
Please answer these questions before submitting a bug report.

### What version of gRPC are you using?
      artifact = 'io.grpc:protoc-gen-grpc-java:1.24.0'

### What operating system (Linux, Windows,...) and version?
       MacOS Mojave : v10.14.6
### What did you expect to see?
      Should compile and generate from proto definitions without errors
### What did you see instead?
```
Task :compileKotlin FAILED e : Supertypes of the following classes cannot be resolved. 
Please make sure you have the required dependencies in the classpath: 
class entity.grpc.User.MyAgeCase, unresolved supertypes: InternalOneOfEnum
```

### Steps to reproduce the bug
    message User {
        oneof my_age {
           string age = 26;
        string age_baz = 29;
       }
    }
Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

##### Gradle
```
buildscript {
    ext.kotlin_version = '1.3.50'

    repositories {
        mavenCentral()
    }

    dependencies {
        classpath 'com.google.protobuf:protobuf-gradle-plugin:0.8.8'
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
    }
}

plugins {
    id 'idea'
    id 'java'
    id "org.flywaydb.flyway" version "6.0.4"
    id 'org.jetbrains.kotlin.jvm' version '1.3.41'
    id 'com.google.protobuf' version '0.8.8'
}

protobuf {
    protoc {
        artifact = "com.google.protobuf:protoc:3.10.0"
    }
    plugins {
        grpc {
            artifact = 'io.grpc:protoc-gen-grpc-java:1.24.0'
        }
    }
    generateProtoTasks {
        all()*.plugins {
            grpc {}
        }
    }

    generatedFilesBaseDir = "$projectDir/build/generated/"
}

sourceCompatibility = 12

repositories {
    mavenCentral()
    maven { url "https://jitpack.io" }
    jcenter()
}

dependencies {
    //Logging
    implementation group: 'org.slf4j', name: 'slf4j-api', version: '1.7.7'
    implementation group: 'org.slf4j', name: 'slf4j-log4j12', version: '1.7.7'
    implementation group: 'javax.annotation', name: 'javax.annotation-api', version: '1.3.2'

    //Postgres Database
    implementation group: 'org.postgresql', name: 'postgresql', version: '42.2.8'

    //Akka
    implementation group: 'com.typesafe.akka', name: 'akka-http_2.12',   version: '10.1.10'
    implementation group: 'com.typesafe.akka', name: 'akka-http2-support_2.12', version: '10.1.10'

    //gRPC - Java
    implementation 'io.grpc:grpc-netty-shaded:1.24.0'
    implementation 'io.grpc:grpc-protobuf:1.24.0'
    implementation 'io.grpc:grpc-stub:1.24.0'
    implementation 'com.google.protobuf:protobuf-java-util:3.9.2'


    //Guice - Google
    implementation group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.10.0'
    implementation group: 'com.google.inject', name: 'guice', version: '4.2.2'
    implementation group: 'com.google.inject.extensions', name: 'guice-assistedinject', version: '4.2.2'

    //Kotlin
    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8"

    //-------------- Test dependencies----------------------------
    testImplementation group: 'junit', name: 'junit', version: '4.12'
    testImplementation "io.mockk:mockk:1.9.3"
    testImplementation("org.assertj:assertj-core:3.11.1")
    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.4.0'
    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.4.0'
    //-------------------------------------------------------------
    implementation 'org.jetbrains.exposed:exposed:0.16.1'
    implementation 'org.hibernate:hibernate-validator:6.0.17.Final'

}


flyway {
    url = 'jdbc:postgresql://localhost:5432/user_management'
    user = 'postgres'
    password = ''
    placeholders = ['keyABC': 'valueXYZ']
}

test{
    useJUnitPlatform()
    testLogging {
        events "passed", "skipped", "failed"
    }
}

compileKotlin {
    kotlinOptions.jvmTarget = "12"
}
compileTestKotlin {
    kotlinOptions.jvmTarget = "12"
}

```

**Comment by sanjaypujare** [comment]
Your protobuf:protoc dependency shows `com.google.protobuf:protoc:3.10.0` but your protobuf-java-util is `com.google.protobuf:protobuf-java-util:3.9.2` . Can you try 3.10.0 for the latter as well?

**Comment by donpaul120** [comment]
@sanjaypujare Oh wow.. That worked. Thanks Very much.
I updated `com.google.protobuf:protobuf-java-util:3.9.2` to  `com.google.protobuf:protobuf-java-util:3.10.0`
Thanks!!!

**Comment by mmoehring** [comment]
System: MacOS Mojave : v10.14.6, gradle 5.6.3

For me there is still an issue with com.google.protobuf:protobuf-java:3.10.0 as version 3.10.0 only gets pulled in when there is a dependency to protobuf-java-util.
If there is no dependency to protobuf-java-util the old version 3.9.0 of protobuf-java gets pulled in which leads to the compile error.

excerpt from build.gradle:
`````
plugins {
    id "com.google.protobuf" version "0.8.8"
}

apply plugin: "com.google.protobuf"



sourceCompatibility = 12

final String grpcv = "1.24.0"
final String protocv = "3.10.0"

dependencies {
    ...
    implementation "com.google.guava:guava"
    implementation "io.grpc:grpc-protobuf:" + grpcv
    implementation "io.grpc:grpc-core:" + grpcv
    implementation "io.grpc:grpc-stub:" + grpcv
    implementation "io.grpc:grpc-netty-shaded:" + grpcv
    ...
}

protobuf {
    generatedFilesBaseDir = "$projectDir/src/"
    protoc {
        artifact = "com.google.protobuf:protoc:" + protocv
    }
    plugins {
        grpc {
            artifact = "io.grpc:protoc-gen-grpc-java:" + grpcv
        }
    }
    generateProtoTasks {
        all()*.plugins {
            grpc {
                outputSubDir = "java"
            }
        }
    }
}

`````

dependencies **without** dependency to "com.google.protobuf:protobuf-java-util:3.10.0":
````
protobuf
No dependencies

protobufToolsLocator_grpc
\--- io.grpc:protoc-gen-grpc-java:1.24.0

protobufToolsLocator_protoc
\--- com.google.protobuf:protoc:3.10.0

+--- io.grpc:grpc-protobuf:1.24.0
|    +--- io.grpc:grpc-api:1.24.0
|    |    +--- io.grpc:grpc-context:1.24.0
|    |    +--- com.google.errorprone:error_prone_annotations:2.3.2
|    |    +--- com.google.code.findbugs:jsr305:3.0.2
|    |    +--- org.codehaus.mojo:animal-sniffer-annotations:1.17 -> 1.18
|    |    \--- com.google.guava:guava:26.0-android -> 28.1-jre (*)
|    +--- com.google.protobuf:protobuf-java:3.9.0
|    +--- com.google.guava:guava:26.0-android -> 28.1-jre (*)
|    +--- com.google.api.grpc:proto-google-common-protos:1.12.0
|    \--- io.grpc:grpc-protobuf-lite:1.24.0
|         +--- io.grpc:grpc-api:1.24.0 (*)
|         \--- com.google.guava:guava:26.0-android -> 28.1-jre (*)

````

dependencies **with** dependency to "com.google.protobuf:protobuf-java-util:3.10.0":
(protobuf-java:3.9.0 gets replaced with 3.10.0):
````
protobuf
No dependencies

protobufToolsLocator_grpc
\--- io.grpc:protoc-gen-grpc-java:1.24.0

protobufToolsLocator_protoc
\--- com.google.protobuf:protoc:3.10.0

+--- com.google.protobuf:protobuf-java-util:3.10.0
|    +--- com.google.protobuf:protobuf-java:3.10.0
|    +--- com.google.guava:guava:28.0-android -> 28.1-jre (*)
|    +--- com.google.errorprone:error_prone_annotations:2.3.2
|    \--- com.google.code.gson:gson:2.8.5 -> 2.8.6
+--- io.grpc:grpc-protobuf:1.24.0
|    +--- io.grpc:grpc-api:1.24.0
|    |    +--- io.grpc:grpc-context:1.24.0
|    |    +--- com.google.errorprone:error_prone_annotations:2.3.2
|    |    +--- com.google.code.findbugs:jsr305:3.0.2
|    |    +--- org.codehaus.mojo:animal-sniffer-annotations:1.17 -> 1.18
|    |    \--- com.google.guava:guava:26.0-android -> 28.1-jre (*)
|    +--- com.google.protobuf:protobuf-java:3.9.0 -> 3.10.0
|    +--- com.google.guava:guava:26.0-android -> 28.1-jre (*)
|    +--- com.google.api.grpc:proto-google-common-protos:1.12.0
|    \--- io.grpc:grpc-protobuf-lite:1.24.0
|         +--- io.grpc:grpc-api:1.24.0 (*)
|         \--- com.google.guava:guava:26.0-android -> 28.1-jre (*)

````

**Comment by sanjaypujare** [comment]
@mmoehring one way to fix your issue is to keep your dependency on `com.google.protobuf:protoc` to 3.9.0 (or 3.9.2?) until `io.grpc:grpc-protobuf` dependency is changed to 3.10.0

**Comment by mmoehring** [comment]
@sanjaypujare Sure, there is a workaround, I'm just unsure what kind of issue this is. The documentation (https://github.com/grpc/grpc-java/commit/079219bbc9c46f3bd99e7cce94429efb2079ff09) states the combination which does not work for me.
So is this a general issue, like, the version is really set to 3.9 in `io.grpc:grpc-protobuf`or is this a problem in the dependency resolution on my system only?

**Comment by sanjaypujare** [comment]
My cursory analysis led to this commit in protobuf https://github.com/protocolbuffers/protobuf/commit/4f02f056b5cea99052bfdfb6698afe47a3cf2964 . This change requires protobuf-java to be on 3.10 if you use protoc 3.10 because of the introduction of [InternalOneOfEnum](https://github.com/protocolbuffers/protobuf/commit/4f02f056b5cea99052bfdfb6698afe47a3cf2964#diff-7ce8586634229ac72ef248b21422bcd0R150). Let me see if this can be fixed in grpc-java release.

**Comment by sanjaypujare** [comment]
@mmoehring looking at the last release 1.24.x I see that it used [3.9.0 for both protobuf and protoc](https://github.com/grpc/grpc-java/blob/v1.24.x/build.gradle#L112) and as long as you stick to 3.9.0 in your application it should be okay. But you used 3.10.0 for protoc. Is that because of some recommendation somewhere?

**Comment by mmoehring** [comment]
@sanjaypujare Thank you for your analysis. In the last two years I took the combination of versions for the build.gradle from the `README.md`, paragraph "Generated Code" (https://github.com/grpc/grpc-java/blob/master/README.md#generated-code). which  right now lists the combination I tried, `com.google.protobuf:protoc:3.10.0"` + `'io.grpc:protoc-gen-grpc-java:1.24.0'`.
This worked for me until now so my assumption was that the README would always document a working combination.

**Comment by sanjaypujare** [comment]
@mmoehring this commit https://github.com/grpc/grpc-java/commit/079219bbc9c46f3bd99e7cce94429efb2079ff09 made the change everywhere consistently. However you used the README.md from master (latest) but the released bits from 1.24.x and that will have this problem.

**Comment by mmoehring** [comment]
@sanjaypujare Yes, that's right. Thantks for your help!
